### PR TITLE
Port SUSE OpenStack Cloud to aarch64

### PR DESCRIPTION
**Pull request by dirkmueller**
Various needed features for an SUSE OpenStack Cloud demo on AArch64

### CODE REVIEW

File: chef/data_bags/crowbar/template-nova.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
       "trusted_flavors": false,
       "libvirt_type": "kvm",
       "use_novnc": true,
+      "use_serial": false,
```
Comment (vuntz) [inline chef/data_bags/crowbar/template-nova.json]: Any reason to not always enable it? Or to not default to `true`?

Comment (vuntz) [inline chef/data_bags/crowbar/template-nova.json]: (btw, you'll need a schema migration)

File: chef/cookbooks/nova/recipes/config.rb [hunk 2]
```diff
@@ -302,11 +317,13 @@
             glance_server_insecure: glance_server_insecure || keystone_settings["insecure"],
             metadata_bind_address: metadata_bind_address,
             vnc_enabled: node[:kernel][:machine] != "aarch64",
+            serial_enabled: node[:kernel][:machine] == "aarch64",
```
Comment (vuntz) [inline chef/cookbooks/nova/recipes/config.rb]: I guess you need to be consistent for the condition, and use `node[:nova][:use_serial]` here.

**Comment by dguitarbite** [comment]
There are some security concerns from my side. Could we configure a VNC password and have it as a configuration option for the same?

File: chef/data_bags/crowbar/template-nova.json (hunk 1 above)
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-nova.json]: yes, its wip for a reason ;)

**Comment by dirkmueller** [comment]
@dguitarbite you mean the iptables blocking thing that you implemented for vnc? otherwise I don't understand what you mean by security concerns. the serial console proxy is behind keystone..

**Comment by dguitarbite** [comment]
@dirkmueller Yes, the code I wrote using iptables for restricting serial console (not just the proxy) ports from the Compute Nodes (libvirt hosts) access. Either we set the password in the ``/etc/libvirt/qemu.conf`` and then make the serial console inject this password to connect or use something like my iptables code (strongly suggest not to use my code ...).

File: chef/cookbooks/nova/recipes/config.rb [hunk 3]
```diff
@@ -285,6 +299,7 @@
             bind_port_objectstore: bind_port_objectstore,
             bind_port_novncproxy: bind_port_novncproxy,
             bind_port_xvpvncproxy: bind_port_xvpvncproxy,
+            bind_port_serialproxy: bind_port_xvpvncproxy,
```
Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/config.rb]: Why is the variable set to ``bind_port_xvpvncproxy`` here where as earlier it was set to another value. I may be missing something here, just skimming through the code.

File: chef/cookbooks/nova/recipes/config.rb (hunk 2 above)
Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/config.rb]: I recommend you to change either the commit message or make a short comment explaining why ``non-x86`` is being used in the commit message where as here the conditional is just checking ``aarch64``. 

``aarch64`` ⊂ ``!x86`` is not true the other way around.

File: chef/data_bags/crowbar/template-nova.json (hunk 1 above)
Comment (dguitarbite) [inline chef/data_bags/crowbar/template-nova.json]: @dirkmueller , I could partially confirm that NoVNC is a dumb server limited by the dumbness of [websockify](https://github.com/kanaka/websockify). Keeping this always on should not be necessary, opening more ports additionally opens up more potential security issues.

@vuntz Please create an additional (noise?) followup card if required for addressing the security concern's w.r.t. enabling ``serial console``.

**Comment by dguitarbite** [comment]
@dirkmueller Small nitpick, is there a possibility to change the 'non-x86' part to more appropriate one in the commit message?

**Comment by toabctl** [comment]
@dirkmueller commit msg needs ```$component: ...``` prefix

**Comment by dirkmueller** [comment]
@toabctl done

File: chef/cookbooks/nova/recipes/config.rb [hunk 4]
```diff
@@ -320,12 +336,14 @@
...
             metadata_bind_address: metadata_bind_address,
-            vnc_enabled: node[:kernel][:machine] != "aarch64",
+            vnc_enabled: node[:nova][:use_novnc],
+            serial_enabled: node[:nova][:use_serial],
             vendordata_jsonfile: vendordata_jsonfile,
             vncproxy_public_host: vncproxy_public_host,
             vncproxy_ssl_enabled: api[:nova][:novnc][:ssl][:enabled],
             vncproxy_cert_file: api_novnc_ssl_certfile,
             vncproxy_key_file: api_novnc_ssl_keyfile,
+            serialproxy_public_host: serialproxy_public_host,
```
Comment (toabctl) [inline chef/cookbooks/nova/recipes/config.rb]: add a var ```serialproxy_ssl_enabled: node[:nova][:serial][:ssl][:enabled]``` and use it later like we do with ```vncproxy_ssl_enabled``` ?

Comment (toabctl) [inline chef/cookbooks/nova/recipes/config.rb]: Also what about ```serialproxy_cert_file``` and ```serialproxy_key_file``` ? you have theses vars in the .schema file

Comment (dirkmueller) [inline chef/cookbooks/nova/recipes/config.rb]: I have removed  ```serialproxy_cert_file``` and ```serialproxy_key_file```

**Comment by dirkmueller** [comment]
@dguitarbite so I looked a bit into this. it seems there is no way to secure serial over ssl or shared secret as far as I can see.

**Comment by dguitarbite** [comment]
@dirkmueller I guess that is the reason for isolating this traffic over a
protected dedicated private network. Solved over architectural design.

+1

**Comment by dguitarbite** [comment]
Hope that is considered or remembered in case of potential request for the
same in future.

**Pull request merged**